keps/sig-network/3705-cloud-node-ips/README.md

```diff
@@ -477,31 +543,46 @@ issues.
 
 ###### Does enabling the feature change any default behavior?
 
-No
+Kubelet will begin setting two annotations on each node rather than
+one. However, cloud providers will still prefer the old annotation and
+will only read the new annotation if the old one isn't there.
+
+So assuming no bugs, the only visible change after (just) enabling the
+feature is the duplicate annotation.
 
 ###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
 
 Yes, as long as you also roll back the kubelet configuration to no
 longer use the new feature.
 
+One edge case: if a user who was not previously setting `--node-ip`
```

Comment:
If this is gated, shouldn't it be:

1.27, gate is alpha, default off
 * if `--node-ip` is unset, delete both annotations
 * if `--node-ip` is set, and gate disabled, set old annotation
 * if `--node-ip` is set, and gate enabled, set both annotations

1.28, gate is beta, default on
 * (same as 1.27)

1.29, gate is GA, default on
 * if `--node-ip` is unset, delete both annotations
 * if `--node-ip` is set, set both annotations

1.3x
 * if `--node-ip` is unset, delete both annotations
 * if `--node-ip` is set, set new annotation

1.3y
 * if `--node-ip` is unset, delete new annotation
 * if `--node-ip` is set, set new annotation

There's no major rush for 3x and even less for 3y - I think?

Reply:
How is 1.3x different from 1.3y ?


+1 for the rest part of the comment
That still leaves the corner case described by Dan, but only for 1.27->1.26 downgrade.
Given that the feature will be Alpha in 1.27, I think the risk is worth taking and not spreading the rollout over additional few releases.

Reply:
> If this is gated, shouldn't it be:
> 
> 1.27, gate is alpha, default off
> 
> - if --node-ip is unset [and gate is disabled], delete both annotations

Oh, yeah, we should do that... I was just ignoring the new annotation entirely when the gate is disabled.

FWIW, that prevents the bug when rolling back from beta to alpha, but it doesn't prevent it when rolling back from alpha-with-feature-gate to pre-alpha. But I think it's edge-case-y enough to be tolerable for people testing the feature in alpha, especially since it only manifests while the cluster is version-skewed, and the "bug" is just that you continue getting the pre-downgrade behavior until you've downgraded both kubelet and cloud-provider to 1.26. (And, if we use the "delete both even with the feature gate is unset" behavior, then you can avoid the edge case by rolling back the kubelet configuration to its previous state before you downgrade the cluster. Which is probably a good idea in general when testing alpha features?)

Reply:
> it doesn't prevent it when rolling back from alpha-with-feature-gate to pre-alpha

In general, we don't (can't) support that.  Consider what happens for a field saved in etcd - if you go to a pre-alpha release it becomes unknown and will get lost on the next write (since it's not known to the internal type).

So don't worry about that.